### ISSUE TITLE
Improvement: quick contacts manage window

### ISSUE THREAD

**Issue by Soulsuke** [emesene/emesene#634]
I've been using emesene2 for a while, and I really miss this emesene feature: the options window where you could see a list of your contacts, and you could check whenever they had you on their list, they were blocked and so on, it really made managing contacts easier (especially the blocked ones, or whenever you had to clean old contacts).

Since I don't know anywhere else where I could ask this... You've stated that changing the multichat rotating avatar with a normal list of the avatars of the participants was out of question (for totally unknown reasons), but someone said that it should be an easy thing to add as a plugin or extension... Now, is someone actually working at it?
I actually want to learn python, so where could i find some documentation about creating emesene2's extensions and plugins?

**Comment by Mystro256** [comment]
https://github.com/emesene/emesene/wiki/Plugin-How-To

http://emesene.github.com/emesene-docs/

and of course: http://wiki.python.org/moin/BeginnersGuide

You can also take a look at the existing plugins too: https://github.com/emesene/emesene-supported-plugins

**Comment by Soulsuke** [comment]
Thanks for the reply!
https://github.com/emesene/emesene-supported-plugins <<< I didn't knew of that, I'll definetly use it :D

https://github.com/emesene/emesene/wiki/Plugin-How-To <<< Thanks a lot! This does helps at least to start writing something from scratches!

http://emesene.github.com/emesene-docs/ <<< Uuuuuh.... I dunno if I'm doing something wrong, but all the pages are either empty or filled with non-technical informations... Is there something like an UML table, or is there some documentation about the files, classes and methods? Or do I have to read through all the code myself?

**Comment by Mystro256** [comment]
Sorry try this instead: http://marianoguerra.com.ar/emesene/api/

The Main page is blank but the Namespaces and Classes pages have some info that might help (it's the best I could find at the momment)

**Comment by Mystro256** [comment]
There is also this: http://marianoguerra.com.ar/emesene/docs but it seems like a copy of the github link I sent up above

**Comment by Soulsuke** [comment]
http://marianoguerra.com.ar/emesene/api/ <<< This one seems to have the info I'm looking for, thanks!

**Comment by Mystro256** [comment]
No problem

**Comment by c10ud** [comment]
quite easy to do with gtk: once it's done just make a module for preference window and hook signals/callbacks from e3..

basically it's 2 liststores with contacts' emails and a bunch of buttons for moving them from ALLOW LIST to BLOCK LIST and vice-versa. i don't remember if any other msn-specific thing is involved, anyway just take a look at emesene1 or wlm for what's needed 

marking this as easy even if it's not _that_ easy, but you got a starting point!

**Comment by manuargue** [comment]
Hi. Somebody is working on it??

**Comment by c10ud** [comment]
nope

**Comment by manuargue** [comment]
ok.. the idea is make it a plugin or a built-in feature?

**Comment by c10ud** [comment]
built in feature inside the Live Messenger tab in Preferences :) along with the "fix profile" stuff

**Comment by manuargue** [comment]
ok.. I can make a frame below the fix profile stuff, containing the two list and buttons to manage contacts. do you like it?

**Comment by c10ud** [comment]
yep that's exactly the idea, similar to what emesene1 had

**Comment by manuargue** [comment]
I already finished to implement this feature.. but I have some troubles with my git repo :)
emesene repo is ahead three commits for my repo.. I did a rebase but nothing happened. Then a (hard) reset but nothing happened.. I'm new with git and I dont know how to fix this.. can u help me?

thanks!

**Comment by c10ud** [comment]
answered in the google group!

**Comment by manuargue** [comment]
something seems to not work properly..
when I login with my msn account, the attribute 'reverse' of the ContactManager instance is empty, while it shoudn't be (I tested it with emesene 1.6 and seems to work properly).. So the feature is working, but apparently there is a problem behind.

**Comment by c10ud** [comment]
the REVERSE flag isn't handled properly by the servers either (yes,
really), so it could be currently not implemented. I am currently not
at home so i'm not sure about the implementation, but i'm 100% sure of
the problems given by the reverse flag: some contacts which should
have it, simply don't -- unknown reasons, and that's directly from the
data the servers give us. My advice is to leave that flag alone for
now, we'll see if something changes, but it's like that since...years
i guess (?)

2011/8/21 manuargue
reply@reply.github.com:

**Comment by c10ud** [comment]
ok, now only the coloring missing and we're done with this issue i guess (?)

**Comment by manuargue** [comment]
yep.. but I don't know where to look for this problem.. maybe create a new issue for the reverse problem?

**Comment by Sbte** [comment]
I don't see any code set self.reverse, so how is this supposed to work? And what is self.reverse for? I don't find the name very descriptive. That doesn't seem to be your mistake though.

**Comment by manuargue** [comment]
I think the reverse dictionary has the contacts that are not in your contact list, but they do have you (at least it was so in emesene1).

I didn't fill this dictionary with the reverse contacts because it must be done outside this class..

**Comment by Sbte** [comment]
The blocked list doesn't seem to work either btw?

**Comment by manuargue** [comment]
I tested it with my msn account and block,unblock and delete works fine to me. And it should work because uses the Session methods to do this job.

**Comment by Sbte** [comment]
I had no one in my blocked list when I checked the widget. I did have at least one contact blocked though.

**Comment by c10ud** [comment]
sure or "maybe"?

**Comment by c10ud** [comment]
```
        if not (papyon.profile.Membership.FORWARD & contact.memberships):
            # This skips contacts that are not in our contact list
            # but are still in the Live Address Book
            continue
```

**Comment by manuargue** [comment]
Sbte:
I think I know where is the problem.. The lists are filled in the class constructor, and it occurs once. If you block/unblock/delete a contact outside the widget, this has effect on the lists.

I'm not at home now, but later I can try to fix that.

**Comment by Sbte** [comment]
Yep, that's my problem. Just figured it out myself too.

**Comment by manuargue** [comment]
2011/8/22 c10ud
reply@reply.github.com:

> if not (papyon.profile.Membership.FORWARD & contact.memberships):
>                # This skips contacts that are not in our contact list
>                # but are still in the Live Address Book
>                continue
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/emesene/emesene/issues/634#issuecomment-1871054

is this code right?

if not (papyon.profile.Membership.FORWARD & contact.memberships):
    # This skips contacts that are not in our contact list
    # but are still in the Live Address Book
    tmp_cont = e3.base.Contact(contact.account, contact.id,
        contact.display_name, contact.personal_message,
        STATUS_PAPY_TO_E3[contact.presence],
        contact.display_name,
        (papyon.profile.Membership.BLOCK & contact.memberships))
    self.session.contacts.reverse[contact.account] = tmp_cont
    continue

**Comment by Sbte** [comment]
If it works it probably is :P

**Comment by c10ud** [comment]
the code is right: but we're not saving in e3 Contacts the different memberships! we're only setting the blocked flag.

also the code i pasted is right for the current situation but at the same time it's not entirely right
- it's right because with _add_contact we're adding to the GUI contact list only the contacts with the FORWARD flag, which means only the contacts we have in our contact list
- it's NOT right because (i think?) your implementation of the allow/block list relies upon the list of the e3 Contact. And that list doesn't contain contacts without the FORWARD flag. But you could have a contact with only the BLOCKED flag, and that contact wouldn't be showed (in the contact list that would be right, but in the privacy tab it's not!)

so this parts need a bit of rework... :(

**Comment by c10ud** [comment]
a quick workaround could be adding to papylib session some methods

e.g. PSEUDOCODE
get_allow
return for contact in papyon addressbook if contact.memberships & membership.ALLOW
get_blocked
for contact in papyon addressbook if contact.memberships & membership.BLOCKED
get_forward
for contact in papyon addressbook if contact.memberships & membership.FORWARD

with these methods we bypass the list that's useful for GUI contactlist only and manage these three lists within the privacy tab.

OR

a single method that returns a dict { account:memberships } for the privacy tab, obviously this method would be in papylib's session

what do you think?

**Comment by manuargue** [comment]
But in the code I submitted, the contacts without the FORWARD flag are saved in the reverse dict, and not in the contacts dict that is used by the GUI contact list.
Thus, to get ALL the block/unblock contacts and fill the privacy lists, you only have to check the Contact.blocked flag in both dictionaries, like I did.

I tested it and seems to work fine.. but, if you don't like it, I'll try to do it in your way ;)

greets

**Comment by c10ud** [comment]
well it was just some theorycrafting on why something was supposedly not working, i didn't extensively test it (yet?) :P

**Comment by manuargue** [comment]
Hi, i'm back :)

I implemented this like c10ud said. I added to the papylib session a few methods to get the contacts with the different flags. In this way, the base ContactManager and the papylib worker remain unchanged.

Also, I subscribed the privacy widget to some signals, so the actions made outside this widget affect it.
The only thing left to "tune" is the update of the lists. When the condition of a contact change (block/unblock/etc), both models are cleared and refilled with the new contacts :) it's a bit dirty..

I guess tomorrow i'll be pushing the changes.
